### PR TITLE
feat: first notification

### PR DESCRIPTION
**Pull request by sshanzel**
### Describe what this PR does
- Notifications context to be included in the boot provider
- Notifications page showing the enable settings banner
- First notification
- Generic component to handle notifications

image

@rebelchris can you help me upload the asset in the Cloudinary? It is about the preview of notification. Link to the Sketch can be found below the screen shot.

image

https://www.sketch.com/s/2867ec63-0966-4647-b1d7-b6fe696ba251/a/m1Q273L#Inspect

Preview:
image

## Events

Did you introduce any new tracking events?
Don't forget to update the [Analytics Taxonomy sheet](https://docs.google.com/spreadsheets/d/18Lv7zXges9QfVX5VYL1a-Hyl0e1sQ3sLr0OK8YZWKXI/edit#gid=0)

Log the new/changed events below:

| Type   | event_name  | value |
|--------|-------------|-------|
| Change/New | event name  | extra: { ... } |

### On those affected packages:
- [ ] Have you done sanity checks in the webapp?
- [ ] Have you done sanity checks in the extension?
- [ ] Does this not break anything in companion?

### Did you test the modified components media queries?
- [ ] MobileL (420px)
- [ ] Tablet (656px)
- [ ] Laptop (1020px)

#### Did you test on actual mobile devices?
- [ ] iOS (Chrome and Safari)
- [ ] Android

WT-702 #done

### CODE REVIEW

File: packages/shared/src/lib/styling.ts [hunk 1]
```diff
@@ -0,0 +1,16 @@
+import classNames from 'classnames';
+
+const fadedBackground =
```
Comment (rebelchris) [inline packages/shared/src/lib/styling.ts:3]: This seems like a lot of overhead I think we can simply get a new color match preferably?
I have the same issue now with feed filters and think we should rather ask design for a solid color?

File: packages/shared/src/components/notifications/NotificationItem.tsx [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+  isUnread,
+  description,
+  avatars,
+  attachments,
+}: NotificationItemProps): ReactElement {
+  const DOMPurify = useRef<DOMPurify.DOMPurifyI>();
+
+  useEffect(() => {
+    DOMPurify.current = createDOMPurify(window);
+  }, []);
```
Comment (rebelchris) [inline packages/shared/src/components/notifications/NotificationItem.tsx:33]: We already use this somewhere correct?
Perhaps we can convert it into a hook (just for clean code)

File: packages/shared/src/components/notifications/NotificationItemAttachment.tsx [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+function NotificationItemAttachment({
+  image,
+  title,
+}: NotificationItemAttachmentProps): ReactElement {
+  return (
+    <div className="flex flex-row items-center p-4 mt-2 rounded-16 border border-theme-divider-tertiary">
+      <img
+        className="object-cover w-24 h-16 rounded-16"
+        src={image}
+        alt="Attachment preview"
```
Comment (rebelchris) [inline packages/shared/src/components/notifications/NotificationItemAttachment.tsx:17]: This should likely be the title rather?

File: packages/shared/src/components/notifications/NotificationItemAvatar.tsx [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+  type,
+  image,
+  name,
+  referenceId,
+}: NotificationItemAvatarProps): ReactElement {
+  if (type === NotificationAvatarType.Source) {
+    return <SourceButton source={{ id: referenceId, name, image }} />;
+  }
+
+  return <ProfilePicture user={{ image, id: referenceId }} />;
```
Comment (rebelchris) [inline packages/shared/src/components/notifications/NotificationItemAvatar.tsx:28]: Generic question:

Should we maybe hard cast a type === here as well?
In case we introduce a new one as we go, rather have that render as empty?

File: packages/shared/src/lib/styling.ts (hunk 1 above)
Comment (rebelchris) [reply packages/shared/src/lib/styling.ts:3]: To iterate, perhaps it's better to add a new scope of colors "faded" and add all the new faded colors?
But suggestion is to ask design for all of them :)

**Review by rebelchris — CHANGES_REQUESTED** [review]
Love it!

Really nice setup with the generic notification.
Have some suggestions for the backgrounds though

File: packages/shared/src/components/utilities.tsx [hunk 5]
```diff
@@ -118,11 +119,14 @@ export const pageBorders =
   'laptop:border-r laptop:border-l border-theme-divider-tertiary';
 const pagePaddings = 'px-4 tablet:px-8';
 
+export const pageContainerClassnames = classNames(
```
Comment (el) [inline packages/shared/src/components/utilities.tsx:122]: pageContainerClassNames

File: packages/shared/src/lib/styling.ts [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+  'relative before:absolute before:inset-0 before:w-full before:h-full before:opacity-24';
+
+export const getFadedBackground = (className: string): string => {
+  const isBefore = className.includes('before:bg-theme-');
+
+  if (!isBefore) {
+    throw new Error(
+      'You must need apply the background with a `before:` class',
```
Comment (el) [inline packages/shared/src/lib/styling.ts:11]:
```
'You need to apply the background with a \'before:\' class'
```

File: packages/shared/src/lib/styling.ts [hunk 7]
```diff
@@ -0,0 +1,16 @@
+import classNames from 'classnames';
+
+const fadedBackground =
+  'relative before:absolute before:inset-0 before:w-full before:h-full before:opacity-24';
+
+export const getFadedBackground = (className: string): string => {
```
Comment (el) [inline packages/shared/src/lib/styling.ts:6]: I think a better approach would be something similar to this:
```
enum Color {
  OverlayCabbage: 'overlay-cabbage'
  ...
}
function generateFadedBackground: (color: Color): string {
  return `before:bg-theme-${color}`;
}
```

**Review by el — APPROVED** [review]
Great work, other than small comments, looks perfect 👏

File: packages/shared/src/components/notifications/NotificationItemAvatar.tsx (hunk 4 above)
Comment (sshanzel) [reply packages/shared/src/components/notifications/NotificationItemAvatar.tsx:28]: Very good point. We shall do that.

File: packages/shared/src/lib/styling.ts (hunk 1 above)
Comment (rebelchris) [reply packages/shared/src/lib/styling.ts:3]: Yeah let's rather ask design.

They can give us 8 digit hex colors so we can adjust them in our guidelines rather.
I prefer having a className for it, compared to a function that needs to execute.

File: packages/shared/src/components/notifications/NotificationItem.tsx (hunk 2 above)
Comment (rebelchris) [reply packages/shared/src/components/notifications/NotificationItem.tsx:33]: If it's reused a hook might still be best, but it's not a hard requirement, so leaving it up to you.

Comment (sshanzel) [reply packages/shared/src/components/notifications/NotificationItem.tsx:33]: Cool. Was just checking if you are cool too with creating a hook to abstract this single useState. In my opinion, we should go with your suggestion so I will do that before moving forward. Thank you!

File: packages/shared/src/components/utilities.tsx (hunk 5 above)
Comment (sshanzel) [reply packages/shared/src/components/utilities.tsx:122]: Sharp eyes! 🔭

File: packages/shared/src/lib/styling.ts (hunk 6 above)
Comment (sshanzel) [reply packages/shared/src/lib/styling.ts:11]: The lint gave an error. Let me make the parent a double quotation as the lint suggested. Thanks for mentioning!

**Comment by sshanzel** [comment]
@rebelchris feel free to run a review, kindly disregard the `fadedBackground` thing until we receive a reply from the message I sent to the design team. Thank you!

File: packages/shared/src/components/notifications/NotificationItem.spec.tsx [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+    const img = await screen.findByAltText(
+      `Preview cover of the article: ${attachment.title}`,
+    );
+    expect(img).toHaveAttribute('src', attachment.image);
+  });
+
+  it('should have a title', async () => {
+    const [attachment] = sampleNotification.attachments;
+    render(<NotificationItem {...sampleNotification} />);
+    await screen.findByText(attachment.title);
```
Comment (rebelchris) [inline packages/shared/src/components/notifications/NotificationItem.spec.tsx:53]: You await but don't expect anything here?

File: packages/shared/src/components/notifications/NotificationItem.spec.tsx [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+  it('should display the avatar of the user', async () => {
+    const [, user] = sampleNotification.avatars;
+    render(<NotificationItem {...sampleNotification} />);
+    const img = await screen.findByAltText(`${user.referenceId}'s profile`);
+    expect(img).toHaveAttribute('src', user.image);
+  });
+
+  it('should not display anything if the type is unknown', async () => {
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    const notif = { ...sampleNotification } as any;
```
Comment (rebelchris) [inline packages/shared/src/components/notifications/NotificationItem.spec.tsx:74]: Should it not cast as `Notification` or something?

File: packages/shared/src/components/utilities.tsx [hunk 10]
```diff
@@ -118,11 +119,14 @@ export const pageBorders =
   'laptop:border-r laptop:border-l border-theme-divider-tertiary';
 const pagePaddings = 'px-4 tablet:px-8';
 
+export const pageContainerClassNames = classNames(
+  styles.pageContainer,
```
Comment (rebelchris) [inline packages/shared/src/components/utilities.tsx:123]: Just a question:
Do we still needs these or can we potentially move them to tailwind classes now?

File: packages/shared/src/contexts/NotificationsContext.tsx [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+    if (!user) return 'default';
+
+    const result = await globalThis.window.Notification.requestPermission();
+    setHasPermission(result === 'granted');
+    return result;
+  };
+
+  const data = useMemo(() => {
+    return {
+      unreadCount: 0,
```
Comment (rebelchris) [inline packages/shared/src/contexts/NotificationsContext.tsx:43]: This is going to be another PR?

**Review by rebelchris — CHANGES_REQUESTED** [review]
Some minor questions and leaving it at changes until we have clarity on the faded background 🙏

File: packages/shared/src/components/notifications/NotificationItem.spec.tsx (hunk 8 above)
Comment (sshanzel) [reply packages/shared/src/components/notifications/NotificationItem.spec.tsx:53]: The await will fail if there is no found element.

File: packages/shared/src/components/notifications/NotificationItem.spec.tsx (hunk 9 above)
Comment (sshanzel) [reply packages/shared/src/components/notifications/NotificationItem.spec.tsx:74]: This is the part where the type of notification is not found in our system. I used any so we can put values outside of the current `enum NotificationType`.

File: packages/shared/src/components/utilities.tsx (hunk 10 above)
Comment (sshanzel) [reply packages/shared/src/components/utilities.tsx:123]: I just checked they are pretty simple, we can probably convert them, though in a separate PR as it affects a lot of things and requires rigorous manual testing.

File: packages/shared/src/contexts/NotificationsContext.tsx (hunk 11 above)
Comment (sshanzel) [reply packages/shared/src/contexts/NotificationsContext.tsx:43]: Yes, the other PR fixes this to come from the props.

File: packages/shared/src/lib/styling.ts (hunk 1 above)
Comment (sshanzel) [reply packages/shared/src/lib/styling.ts:3]: Now that we have concluded the usage of opacity, I think we will have to move forward with the function to generate it until we upgrade our tailwind version.

We probably need to put some comments on it to make sure it gets updated. WDYT?

Comment (rebelchris) [reply packages/shared/src/lib/styling.ts:3]: As far as I was concerned the solution was actually to generate classes for it right?
I feel like we might be misaligned here from what feedback we received let's rather discuss it quickly today.

Comment (sshanzel) [reply packages/shared/src/lib/styling.ts:3]: Let's jump in Gather once you are available.

**Comment by sshanzel** [comment]
@rebelchris I have now applied the discussed classes for the different states.

File: packages/shared/src/components/notifications/NotificationItem.tsx [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+      <NotificationItemAvatar key={avatar.referenceId} {...avatar} />
+    )) ?? [];
+  const hasAvatar = avatarComponents.some((component) => !!component);
+
+  return (
+    <button
+      type="button"
+      className={classNames(
+        'flex flex-row py-4 pl-6 pr-4 hover:bg-theme-hover focus:bg-theme-active',
+        isUnread && 'bg-theme-float',
```
Comment (rebelchris) [inline packages/shared/src/components/notifications/NotificationItem.tsx:39]: What's the fallback for read messages in this case?

**Review by rebelchris — APPROVED** [review]
Amazing!

Sorry for dragging it on so long, but it looks really good now 🥳
